docs/rfcs/032-compute-aux-files.md

```diff
@@ -0,0 +1,112 @@
+# Postgres aux files storage
+
+In the current Neon architecture, compute nodes can only persist data through WAL replication to storage nodes. This method covers most of the data that Postgres persists, but not all. Exceptions include:
+
+* Replication slots
+* Postgres statistics files
+* `pg_stat_statements`` disk state
+* `pg_prewarm` disk state
+
+It may also be beneficial to store observability data, such as compute metrics and page access histograms, in the per-endpoint storage.
+
+## Aux file types
+
+The mentioned files have different requirements regarding durability, exhibit different I/O patterns, vary in expected size, and have different desired behaviors with respect to point-in-time recovery and read-only replicas.
+
+### Replication Slots
+
+Replication slots persist data about the replication receiver on the replication source. In the case of chain replication, these files could be modified by a replica (because it may have another attached replica). The ability to modify this state on a replica is the main reason for this data to be kept in a separate file instead of in the usual relation (replicas can't modify data).
+
+* Stale file: Not allowed, as Postgres might have already pruned WAL or vacuumed old catalog entries (relevant in the case of logical replication).
+* Lost file: Postgres will start, but replication will not be able to proceed.
+* Point-in-time recovery (PITR): Resetting a slot to a previous point in time is meaningful in that it allows replication to continue from that point. However, it's hard to envision a use case where this would be useful, as the replica has to be reset in a similar fashion. Considering that a stale slot will cause WAL and catalog to grow without garbage collection, we decided to delete slots upon branch creation. This decision aligns with Postgres documentation's suggestion to omit `pg_replslot` from backups.
+
+There are several types of files:
+
+* Slot state, `pg_replslot/<slot_name>/state`, fixed size, about 200 bytes; updated on checkpoint.
+* LogicalRewriteHeap, `pg_logical/mappings/map-*`, size varies with the size of the mapping; updated on checkpoint.
+* SnapBuild state, `pg_logical/snapshots/*`, fixed structure plus an array of transaction IDs, size can reach up to kilobytes; updated on checkpoint or with running_xact WAL record.
+* ReplicationOrigin, `pg_logical/replorigin_checkpoint`, fixed size, capped at 10 bytes; updated on checkpoint.
+
+An important point is that these files could be modified on the replica in cases of chain replication or logical decoding at standby. Our approach with WAL logging comes at the expense of not being able to support chain replication and logical decoding at standby.
```

Comment:
With my proposed centralized slot service:

- `pg_replslot/<slot_name>/state`. The primary source of truth for this data is in the centralized slot service. On startup, the slot states are downloaded from the centralized service and written to local state files. Thereafter, they are synchronized periodically.

- `pg_logical/mappings/map-*`. These are WAL-logged and stored in pageservers like other data (we are missing the ingestion of these ATM, but that needs to be fixed). These could be garbage collected early, as soon as the LSN becomes older than the oldest restart LSN, but it's not strictly required to do so, we can keep them around longer. In any case, we need to implement garbage collection for them. In the compute, we can implement lazy download for these like we did for SLRUs.

- `pg_logical/snapshots/*`. Let's store these in the centralized service. When a compute creates a snapshot file, it can upload it to the centralized service. Or do it at checkpoints. It's OK if multiple compute upload the same file, as the contents of a file with the same name should be identical.

- `pg_logical/replorigin_checkpoint`. These need to be handled by the storage like relations and SLRUs.